sdks/python/apache_beam/runners/direct/direct_userstate.py

```diff
@@ -225,14 +225,28 @@ def __init__(self, step_context, dofn, key_coder):
 
     self.cached_states = {}
     self.cached_timers = {}
-
-  def get_timer(self, timer_spec, key, window, timestamp, pane):
+    self.cached_timer_families = {}
+
+  def get_timer(
+      self,
+      timer_spec: userstate.TimerSpec,
+      key,
+      window,
+      timestamp,
+      pane,
+      is_timer_family=False
```

Comment:
I thought about that but it seemed too much code duplication.

Reply:
I don't think it will result in too much code duplication: basically splitting your if branch into 2 functions and create one util function for creating cached_key.
I'm leaning to better readability over saving code duplication.